types/sanctuary/index.d.ts

```diff
@@ -186,98 +179,106 @@ declare namespace Sanctuary {
     //  Array
-    //  TODO: Fantasyland overloads, non-curried versions
+    array(tk: TK): (tk: TK) => (tk: TK) => TK;
+    head(tk: TK): TK;
+    last(tk: TK): TK;
+    tail(tk: TK): TK;
+    init(tk: TK): TK;
+    take(tk: TK): (tk: TK) => TK;
+    drop(tk: TK): (tk: TK) => TK;
+    takeLast(tk: TK): (tk: TK) => TK;
+    dropLast(tk: TK): (tk: TK) => TK;
+    takeWhile(tk: TK): (tk: TK) => TK;
+    dropWhile(tk: TK): (tk: TK) => TK;
+    size(tk: TK): TK;
+    all(tk: TK): (tk: TK) => TK;
+    any(tk: TK): (tk: TK) => TK;
+    none(tk: TK): (tk: TK) => TK;
     append<A>(x: A): {
-      (xs: ReadonlyArray<A>): A[];
+      (xs: ReadonlyArray<A>): Array<A>;
       (xs: Applicative<A>): Applicative<A>;
     };
     prepend<A>(x: A): {
-      (xs: ReadonlyArray<A>): A[];
+      (xs: ReadonlyArray<A>): Array<A>;
       (xs: Applicative<A>): Applicative<A>;
     };
     joinWith(p: string): (q: ReadonlyArray<string>) => string;
     elem<A>(p: A): (q: Foldable<A> | StrMap<A> | ReadonlyArray<A>) => boolean;
     find<A>(p: Predicate<A>): (q: ReadonlyArray<A> | StrMap<A> | Foldable<A>) => Maybe<A>;
-    pluck(key: string): (xs: Functor<any>) => Functor<any>;
-    unfoldr<A, B>(f: Fn<B, Maybe<Pair<A, B>>>): (x: B) => A[];
-    range(from: Integer): (to: Integer) => Integer[];
-    groupBy<A>(f: Fn2<A, A, boolean>): (xs: ReadonlyArray<A>) => A[][];
-    reverse<A>(foldable: ReadonlyArray<A>): A[];
+    foldMap(tk: TK): (tk: TK) => TK;
+    unfoldr<A, B>(f: Fn<B, Maybe<Pair<A, B>>>): (x: B) => Array<A>;
+    range(from: Integer): (to: Integer) => Array<Integer>;
+    groupBy<A>(f: Fn2<A, A, boolean>): (xs: ReadonlyArray<A>) => Array<Array<A>>;
+    reverse<A>(foldable: ReadonlyArray<A>): Array<A>;
     reverse<A>(foldable: Foldable<A>): Foldable<A>;
-    sort<A>(foldable: ReadonlyArray<A>): A[];
+    sort<A>(foldable: ReadonlyArray<A>): Array<A>;
     sort<A>(foldable: Foldable<A>): Foldable<A>;
     sortBy<A>(f: Fn<A, Ord<any>>): {
-      (foldable: ReadonlyArray<A>): A[];
+      (foldable: ReadonlyArray<A>): Array<A>;
       (foldable: Foldable<A>): Foldable<A>;
     };
+    zip<A, B>(as: ReadonlyArray<A>): (bs: ReadonlyArray<B>) => Array<Pair<A, B>>;
+    zipWith<A, B, C>(fn: Fn2<A, B, C>): (as: ReadonlyArray<A>) => (bs: ReadonlyArray<B>) => Array<C>;
     //  Object
     prop(p: string): (q: any) => any;
     props(p: ReadonlyArray<string>): (q: any) => any;
     get(p: Predicate<any>): (q: string) => (r: any) => Maybe<any>;
     gets(p: Predicate<any>): (q: ReadonlyArray<string>) => (r: any) => Maybe<any>;
     //  StrMap
-    keys(p: StrMap<any>): string[];
-    values<A>(p: StrMap<A>): A[];
-    pairs<A>(p: StrMap<A>): Array<Pair<string, A>>;
+    value(k: string): <A>(m: StrMap<A>) => Maybe<A>;
+    singleton(k: string): <A>(v: A) => StrMap<A>;
+    insert(k: string): <A>(v: A) => (m: StrMap<A>) => StrMap<A>;
+    remove(k: string): <A>(m: StrMap<A>) => StrMap<A>;
+    keys(m: StrMap<any>): Array<string>;
+    values<A>(m: StrMap<A>): Array<A>;
```

Comment:
@vicrac, is there any way to make `S.values ({foo: 1, bar: 2, baz: /XXX/})` a type error?

Reply:
Yes, but it requires explicit type parameter:
```
S.values<number>({ a: 1, bar: 2, baz: 'foo' })
```
It errors out with:
```
Type 'string' is not assignable to type number
```
If type parameter not provided, it's still perfectly valid code and Typescript correctly infers `A` to be `number | string`

Reply:
> If type parameter not provided, it's still perfectly valid code and Typescript correctly infers `A` to be `number | string`

It's valid according to TypeScript but not according to Sanctuary. I suppose TypeScript users will disable Sanctuary's run-time type checking, though, so the current behaviour seems reasonable.

Reply:
Why exactly is it not valid in Sanctuary?

Reply:
Haskell does not support [untagged unions][1]:

```
Prelude> ["foo", True, 42]

<interactive>:1:9: error:
    • Couldn't match expected type ‘[Char]’ with actual type ‘Bool’
    • In the expression: True
      In the expression: ["foo", True, 42]
      In an equation for ‘it’: it = ["foo", True, 42]
```

Sanctuary, which is modelled on Haskell, does not support untagged unions either:

```
> S.I (['foo', true, 42])
Thrown:
TypeError: Type-variable constraint violation

I :: a -> a
     ^
     1

1)  ["foo", true, 42] :: Array ???

Since there is no type of which all the above values are members, the type-variable constraint has been violated.
```

Haskell does support [tagged unions][2] (algebraic data types):

```haskell
Prelude> data FooBarBaz = Foo String | Bar Bool | Baz Int deriving Show

Prelude> [Foo "foo", Bar True, Baz 42]
[Foo "foo",Bar True,Baz 42]
```

Sanctuary supports algebraic data types, and provides several of its own.

Sanctuary was designed to make writing bug-free code in vanilla JavaScript a realistic possibility. In this context, untagged unions are as problematic as `null`: they can lead to run-time type errors if one forgets to include the run-time checks and branching logic necessary to handle each of the cases correctly.


[1]: https://en.wikipedia.org/wiki/Union_type#Untagged_unions
[2]: https://en.wikipedia.org/wiki/Tagged_union